### ISSUE TITLE
Conversations ends up in wrong chat after rebuilding the chat activity

### ISSUE THREAD

**Issue by vanitasvitae** [siacs/Conversations#1969]
#### General information

* **Version:** 1.13.5
* **Device:** Google Nexus 5
* **Android Version:** Android 6.0 CopperheadOS
* **Server name:** self hosted
* **Server software:** ejabberd 16.06
* **Installed server modules:** Stream Managment, CSI, MAM
* **Conversations source:** F-Droid

#### Steps to reproduce

1. Open chat A from notifications (!)
2. Open chat B from chat List (optionally wrote some messages there)
3. Open recent app list and go back to conversations (or trigger activity recreation on another way eg. open chat details and go back)

#### Expected result

What is the expected output? 
App ends up in chat B again.

What do you see instead?
App ends up in chat A

**Comment by licaon-kter** [comment]
If B is latest used I see nothing wrong.

**Comment by vanitasvitae** [comment]
Read again: I actually end up in A.

The problem does not occure if conversations was launched from launcher and never by notification.

**Comment by iNPUTmice** [comment]
Just for my own reference. Usually after receiving a VIEW intent (this is
what happens if you open Conversations via a notification) the view intent
gets invalidated (overwritten) with an empty intent signaling us that we
already used up this intent and don't want to process this further. Somehow
this doesn't seem to happen in this case....

**Comment by licaon-kter** [comment]
Oh, my bad.

**Comment by vanitasvitae** [comment]
Starting Conversations from Notification (Chat A):
´´´´
07-24 00:27:50.706 18591 18613 I conversations: AxolotlService (myaccount@myserver.bla): PreKeyWhisperMessage received, new session ID:1/Optional.of(17)
07-24 00:27:50.751 18591 18613 D conversations: AxolotlService (myaccount@myserver.bla):  Received Message with session fingerprint: SOMELONGFINGERPRINT
07-24 00:27:50.769 18591 18613 D conversations: update unread count to 1
07-24 00:27:50.963 18591 18613 D conversations: myaccount@myserver.bla: successfully fetched vCard avatar for conversations@conference.siacs.eu/Gerc
07-24 00:27:52.311 18591 18613 D conversations: AxolotlService (myaccount@myserver.bla): Bundle XXXXXXXXX in PEP was current
07-24 00:27:54.658 18591 18591 D conversations: app switched into foreground
07-24 00:27:54.658 18591 18591 D conversations: onBackendConnected()
07-24 00:27:54.658 18591 18591 D conversations: onBackendConnected() - stored intent was view_conversations
07-24 00:27:54.659 18591 18591 D conversations: handleViewConversationIntent()
07-24 00:27:54.709 18591 18591 D conversations: update unread count to 0
07-24 00:27:54.716 18591 18591 D conversations: myaccount@myserver.bla: sending read marker to contact@server.bla/conversations
07-24 00:27:54.725 18591 18591 D conversations: onBackendConnected() - overwriting intent
07-24 00:27:55.030 18591 18591 D conversations: load more messages for Contact Name prior to 2016-07-23T22:06:29.670Z
07-24 00:27:55.034 18591 18611 D conversations: myaccount@myserver.bla: running mam query with=contact@server.bla, start=1970-01-01T00:00:00.000Z, end=2016-07-23T22:06:29.670Z
07-24 00:27:55.354 18591 18613 D conversations: myaccount@myserver.bla: finished mam after 0 messages. messages left=false
07-24 00:27:55.380 18591 18591 D conversations: load more messages for Contact Name prior to 2016-07-23T22:06:29.670Z
07-24 00:28:08.240 18591 18613 D conversations: myaccount@myserver.bla: activating grace period till 00:30:32
07-24 00:28:09.022 18591 18613 D conversations: myaccount@myserver.bla: activating grace period till 00:30:33
````
When going from chat B to recent apps and ending up in A again:
````
07-24 00:33:14.931 18591 18591 D conversations: app switched into background
07-24 00:33:15.968 18591 18591 D conversations: app switched into foreground
07-24 00:33:15.968 18591 18591 D conversations: onBackendConnected()
07-24 00:33:15.968 18591 18591 D conversations: onBackendConnected() - stored intent was view_conversations
07-24 00:33:15.968 18591 18591 D conversations: handleViewConversationIntent()
07-24 00:33:15.974 18591 18591 D conversations: onBackendConnected() - overwriting intent
07-24 00:33:16.100 18591 18591 D conversations: load more messages for Contact Name prior to 2016-07-23T22:06:29.670Z
````

**Comment by vanitasvitae** [comment]
And this happens when I start the app from launcher after swiping it away in recent apps:
I start the app, select chat A, select chat B, go to recent apps and back again.
````
07-24 00:41:12.741 20318 20318 D conversations: app switched into foreground
07-24 00:41:12.741 20318 20318 D conversations: onBackendConnected()
07-24 00:41:12.772 20318 20318 D conversations: onBackendConnected() - overwriting intent
07-24 00:41:19.694 20318 20318 D conversations: load more messages for Contact Name prior to 2016-07-23T22:06:29.670Z
07-24 00:41:19.696 20318 20369 D conversations: myaccount@myserver.bla: running mam query with=contact@server.bla, start=1970-01-01T00:00:00.000Z, end=2016-07-23T22:06:29.670Z
07-24 00:41:20.014 20318 20371 D conversations: myaccount@myserver.bla: finished mam after 0 messages. messages left=false
07-24 00:41:20.079 20318 20318 D conversations: load more messages for Contact Name prior to 2016-07-23T22:06:29.670Z
07-24 00:41:27.438 20318 20318 D conversations: app switched into background
07-24 00:41:28.475 20318 20318 D conversations: app switched into foreground
07-24 00:41:28.475 20318 20318 D conversations: onBackendConnected()
07-24 00:41:28.503 20318 20318 D conversations: onBackendConnected() - overwriting intent
````

**Comment by iNPUTmice** [comment]
Hi,

thanks for the log. That helped me a lot to understand what is going  on although I have still no idea why. Somehow the API behaves differently on your phone. 
I have to find a work around but as I can't test this myself we have to do some trail and error together.
I updated that branch again https://github.com/siacs/Conversations/tree/more_logging_for_view_intent with a first test on how to work around this issue. Could you test this for me?

**Comment by vanitasvitae** [comment]
Sure I'll do that sometime this day

**Comment by vanitasvitae** [comment]
Logs of new version:
````
#### Open Conversations from Launcher and open chat A ####
07-24 22:06:02.283 17777 17777 D conversations: app switched into foreground
07-24 22:06:02.284 17777 17777 D conversations: onBackendConnected()
07-24 22:06:02.576 17777 17777 D conversations: load more messages for One Contact prior to 2016-07-24T19:50:43.168Z
07-24 22:06:02.578 17777 17813 D conversations: myaccount@myserver.bla: running mam query with=onecontact@oneserver.bla, start=1970-01-01T00:00:00.000Z, end=2016-07-24T19:50:43.168Z
07-24 22:06:03.495 17777 17815 D conversations: myaccount@myserver.bla: finished mam after 0 messages. messages left=false
07-24 22:06:03.551 17777 17777 D conversations: load more messages for One Contact prior to 2016-07-24T19:50:43.168Z
07-24 22:07:01.368 17777 17777 D conversations: load more messages for One Contact prior to 2016-07-24T19:50:43.168Z
07-24 22:07:04.803 17777 17777 D conversations: app switched into background
07-24 22:07:05.616 17777 17777 D conversations: app switched into foreground
07-24 22:07:05.616 17777 17777 D conversations: onBackendConnected()
07-24 22:07:05.788 17777 17777 D conversations: load more messages for One Contact prior to 2016-07-24T19:50:43.168Z
07-24 22:07:09.374 17777 17777 D conversations: app switched into background
07-24 22:07:10.222 17777 17777 D conversations: app switched into foreground
07-24 22:07:10.222 17777 17777 D conversations: onBackendConnected()
07-24 22:07:32.016 17777 17777 D conversations: app switched into background
#### Get message chat B #####
07-24 22:07:59.195 17777 17815 D conversations: update unread count to 1
07-24 22:08:08.503 17777 17777 D conversations: onNewIntent()
07-24 22:08:08.503 17777 17777 D conversations: onNewIntent(): service was not bound. saving for later
07-24 22:08:08.513 17777 17777 D conversations: app switched into foreground
07-24 22:08:08.513 17777 17777 D conversations: onBackendConnected()
07-24 22:08:08.513 17777 17777 D conversations: onBackendConnected() - stored intent was view_conversations
07-24 22:08:08.513 17777 17777 D conversations: handleViewConversationIntent()
07-24 22:08:08.520 17777 17777 D conversations: update unread count to 0
07-24 22:08:08.521 17777 17777 D conversations: myaccount@myserver.bla: sending read marker to onecontact@oneserver.bla/conversations
07-24 22:08:08.721 17777 17777 D conversations: load more messages for One Contact prior to 2016-07-24T19:50:43.168Z
#### Go to recent apps and back multiple times - no problem ####
07-24 22:08:15.465 17777 17777 D conversations: app switched into background
07-24 22:08:16.477 17777 17777 D conversations: app switched into foreground
07-24 22:08:16.477 17777 17777 D conversations: onBackendConnected()
07-24 22:08:20.003 17777 17777 D conversations: load more messages for One Contact prior to 2016-07-24T19:50:43.168Z
07-24 22:08:23.959 17777 17777 D conversations: app switched into background
07-24 22:08:24.789 17777 17777 D conversations: app switched into foreground
07-24 22:08:24.789 17777 17777 D conversations: onBackendConnected()
07-24 22:08:24.961 17777 17777 D conversations: load more messages for One Contact prior to 2016-07-24T19:50:43.168Z
07-24 22:08:35.951 17777 17777 D conversations: app switched into background
07-24 22:08:39.874 24978 25012 D conversations: restoring roster
07-24 22:08:39.883 24978 24978 D conversations: update unread count to 0
07-24 22:08:39.888 24978 24978 D conversations: schedule next idle ping
#### Get message from B, open Conversations from notification ####
07-24 22:10:16.444 24978 25014 D conversations: update unread count to 1
07-24 22:10:17.105 24978 25014 D conversations: myaccount@myserver.bla: finished mam after 0 messages. messages left=false
07-24 22:10:23.274 24978 24978 D conversations: app switched into foreground
07-24 22:10:23.274 24978 24978 D conversations: onBackendConnected()
07-24 22:10:23.274 24978 24978 D conversations: onBackendConnected() - stored intent was view_conversations
07-24 22:10:23.275 24978 24978 D conversations: handleViewConversationIntent()
07-24 22:10:23.296 24978 24978 D conversations: update unread count to 0
07-24 22:10:23.300 24978 24978 D conversations: myaccount@myserver.bla: sending read marker to anothercontact@myserver.bla/phone
07-24 22:10:23.513 24978 24978 D conversations: load more messages for Another Contact prior to 2016-07-24T20:10:16.399Z
07-24 22:10:37.255 24978 24978 D conversations: app switched into background
07-24 22:10:37.881 24978 24978 D conversations: app switched into foreground
07-24 22:10:37.881 24978 24978 D conversations: onBackendConnected()
07-24 22:10:37.881 24978 24978 D conversations: onBackendConnected() - stored intent was view_conversations
07-24 22:10:37.881 24978 24978 D conversations: handleViewConversationIntent()
#### Open Chat B ####
07-24 22:10:38.039 24978 24978 D conversations: load more messages for Another Contact prior to 2016-07-24T20:10:16.399Z
07-24 22:10:40.075 24978 24978 D conversations: load more messages for One Contact prior to 2016-07-24T19:50:43.168Z
07-24 22:10:40.076 24978 25012 D conversations: myaccount@myserver.bla: running mam query with=onecontact@oneserver.bla, start=1970-01-01T00:00:00.000Z, end=2016-07-24T19:50:43.168Z
07-24 22:10:40.794 24978 25014 D conversations: myaccount@myserver.bla: finished mam after 0 messages. messages left=false
07-24 22:10:40.861 24978 24978 D conversations: load more messages for One Contact prior to 2016-07-24T19:50:43.168Z
#### Recent apps and back - wrong chat ####
07-24 22:10:40.975 24978 24978 D conversations: app switched into background
07-24 22:10:41.731 24978 24978 D conversations: app switched into foreground
07-24 22:10:41.731 24978 24978 D conversations: onBackendConnected()
07-24 22:10:41.731 24978 24978 D conversations: onBackendConnected() - stored intent was view_conversations
07-24 22:10:41.731 24978 24978 D conversations: handleViewConversationIntent()
07-24 22:10:41.850 24978 24978 D conversations: load more messages for Another Contact prior to 2016-07-24T20:10:16.399Z
````

**Comment by iNPUTmice** [comment]
Hi, 

thanks for the log. I think I'm able to reproduce this now.
I think this is a result of an enabled *Don't keep Activities* in the developer Settings. Which might or not might not be enabled by default on Copperhead OS (Can't imagine why they would.) I tested this before with this setting but never extensively enough to actually run into that bug.

Anyway I pushed another commit to that branch ba0bd1ab55847050e1b160e96147ca0a6dbbcb55 that seems to address this issue however I want to run this for a few days myself to make sure there aren't any side effects.

Feel free to check this out and also check if that setting I mentioned is enabled on your device. (Just out of curiosity. Conversations should of course work no matter if that setting is set or not.)
